designs/0100-reduce_sum.md

```diff
@@ -0,0 +1,261 @@
+- Feature Name: Reduce Sum Parallelization
+- Start Date: 2020-03-10
+- RFC PR:
+- Stan Issue:
+
+# Summary
+[summary]: #summary
+
+The reduce_sum function is an attempt to add a parallelization utility that is much easier to use than map-rect without requiring a substantial changes to the autodiff stack.
+
+# Motivation
+[motivation]: #motivation
+
+The goal of reduce_sum is to make it easier for users to parallelize their models by streamlining how arguments are handled, hiding work scheduling, and making it more difficult to program something that will accidentally be inefficient.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+```reduce_sum``` is a tool for parallelizing operations that can be represented as a sum of functions, `g: U -> real`.
+
+For instance, for a sequence of ```x``` values of type ```U```, ```{ x1, x2, ... }```, we might compute the sum:
+
+```g(x1) + g(x2) + ...```
+
+In terms of probabilistic models, comes up when there are N conditionally independent terms in a likelihood. In these cases each conditionally indepedent term can be computed in parallel. If dependencies exist between the terms, then this isn't possible. For instance, in evaluating the log density of a Gaussian process ```reduce_sum``` would not be very useful.
+
+```reduce_sum``` doesn't actually take ```g: U -> real``` as an input argument. Instead it takes ```f: U[] -> real```, where ```f``` computes the partial sum corresponding to the slice of the sequence ```x``` passed in. For instance:
+
+```
+f({ x1, x2, x3 }) = g(x1) + g(x2) + g(x3)
+f({ x1 }) = g(x1)
+f({ x1, x2, x3 }) = f({ x1, x2 }) + f({ x3 })
+```
+
+If the user can write a function ```f: U[] -> real``` to compute the necessary partial sums in the calculation, then we can provide a function to automatically parallelize the calculations (and this is what ```reduce_sum``` is).
+
+If the set of work is represented as an array ```{ x1, x2, x3, ... }```, then mathematically it is possible to rewrite this sum with any combination of partial sums.
+
+For instance, the sum can be written:
+
+1. ```f({ x1, x2, x3, ... })```, summing over all arguments, using one partial sum
+2. ```f({ x1, ..., xM }) + reduce({ x(M + 1), x(M + 2), ...})```, computing the first M terms separately from the rest
+3. ```f({ x1 }) + f({ x2 }) + f({ x3 }) + ...```, computing each term individually and summing them
+
+The first form uses only one partial sum and no parallelization can be done, the second uses two partial sums and so can be parallelized over two workers, and the last can be parallelized over as many workers as there are elements in the array of ```x```s. Depending on how the list is sliced up, it is possible to parallelize this calculation over many workers.
+
+```reduce_sum``` is the tool that will allow us to automatically parallelize this calculation. ```x``` is called the sliced array, because it will be automatically sliced up for the parallelization.
```

Comment:
could we please call this argument `x` ?
if you omit words "is called the sliced array, because" everything is much clearer.